JDBC/OAuthSampleApp/src/main/java/OAuthSampleApp.java

```diff
@@ -70,14 +147,128 @@ private static void printResults(ResultSet rs) throws SQLException {
+                result.append("&");
+            }
+            result.append(URLEncoder.encode(entry.getKey(), StandardCharsets.UTF_8.name()));
+            result.append("=");
+            result.append(URLEncoder.encode(entry.getValue(), StandardCharsets.UTF_8.name()));
+        }
+        return result.toString();
+    }
+
+    // password grant logs into the IDP using credentials in app.config
+    public static void GetTokensByPasswordGrant()  throws Exception {
+    	Map<String, String> formData = new HashMap<String, String>();
+		formData.put("grant_type",   "password");
+		formData.put("client_id",     prop.getProperty("ClientId"));
+		formData.put("client_secret", prop.getProperty("ClientSecret"));
+		formData.put("username",      prop.getProperty("User"));
+		formData.put("password",      prop.getProperty("Password"));
+		GetAndSetTokens(formData);
+    }
 
+    // refresh grant uses the refresh token to get a new access and refresh token
+    public static void GetTokensByRefreshGrant() throws Exception {
+    	Map<String, String> formData = new HashMap<String, String>();
+   		formData.put("grant_type",		"refresh_token");
+		formData.put("client_id",		prop.getProperty("ClientId"));
+		formData.put("client_secret",	prop.getProperty("ClientSecret"));
+		formData.put("refresh_token",	System.getProperty(OAUTH_REFRESH_TOKEN_VAR_STRING));	
+
+		GetAndSetTokens(formData);
+    }
+    
+	private static void GetAndSetTokens(Map<String, String> formData) throws Exception
+    {
+    	try {
+	    	String postOpts = getEncodedParamsString(formData);
+	        byte[] postData = postOpts.getBytes("UTF-8");
+	        int postDataLength = postData.length;
+	        URL url = new URL(prop.getProperty("TokenUrl"));
+	        
+	        HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+	        try {
+	            connection.setDoOutput(true);
+	            connection.setUseCaches(false);
+	            connection.setRequestMethod("POST");
+	            connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
+	            connection.setRequestProperty("Content-Length", Integer.toString(postDataLength));
+	            connection.setRequestProperty("Accept", "application/json");
+				connection.getOutputStream().write(postData);
+	           	BufferedInputStream in = new BufferedInputStream(connection.getInputStream());
+	    	    ByteArrayOutputStream buf = new ByteArrayOutputStream();
+	            for (int result = in.read(); result != -1; result = in.read()) {
+	               	buf.write((byte) result);
+		        }
+	    	    String res = buf.toString("UTF-8");
+	        	JsonElement jElement = new JsonParser().parse(res);
+	            JsonObject jObject = jElement.getAsJsonObject();
+	            
+				// set Tokens as System Properties - new values to access_token and refresh_token
+				System.setProperty(OAUTH_ACCESS_TOKEN_VAR_STRING, jObject.get("access_token").getAsString());
```

Comment:
There should be a null/empty check for the access token and refresh token

Reply:
Added the Null/empty check for the access and refresh token.